packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+import 'dart:ui';
+
+import 'package:flutter/rendering.dart';
+
+import 'basic.dart';
+import 'container.dart';
+import 'framework.dart';
+import 'inherited_theme.dart';
+import 'navigator.dart';
+import 'overlay.dart';
+
+/// [MagnifierController]'s main benefit over holding a raw [OverlayEntry] is that
+/// [MagnifierController] will handle logic around waiting for a magnifier to animate in or out.
+///
+/// If a magnifier chooses to have an entry / exit animation, it should provide the animation
+/// controller to [MagnifierController.animationController]. [MagnifierController] will then drive
+/// the [AnimationController] and wait for it to be complete before removing it from the
+/// [Overlay].
+///
+/// To check the status of the magnifier, see [MagnifierController.shown].
+// TODO(antholeole): This whole paradigm can be removed once portals
+// lands - then the magnifier can be controlled though a widget in the tree.
+// https://github.com/flutter/flutter/pull/105335
+class MagnifierController {
```

Comment:
This seems to be a bit strange, the class is exposing the getter and setter to its `animationController` (along with some other things that also look like implementation details). Is this supposed to be created and used by app developers?

Reply:
Here's the problem / solution that lead to animationController being settable:

A magnifier is told elsewhere to remove itself from the overlay. We can't do `overlayEntry.remove`, since the magnifier might have a exit animation; if we just remove it directly, we will never be able to play that animation. Since the only object that knows the magnifier's exit animation duration is the magnifier, the magnifier gives the MagnifierController an animation controller (hence the setter) and when MagnifierController.hide is called, the magnifier controller awaits on animationController.reverse(), and then removes from the overlay.

Is there another solution to this? I don't want to "leak" animation details like duration to anywhere but the magnifier itself. We can talk about this offline

Reply:
Could that be accomplished using `hide`'s Future?  I read through the code where animationController is passed in and used in Cupertino(Text)Magnifier and it looks like it's more involved than would be possible for just a Future, but from your description in seems like the Future would work.

For overlayEntry I agree it needs to be exposed in order to set the order of the widgets in the Overlay.

Reply:
Talked offline and I was misunderstanding a kind of chicken and egg problem with the Future and the AnimationController.

This reminds me of animated page transitions in Flutter, like animating a route off the screen with something like PageRouteBuilder.transitionsBuilder before removing it.  Since that infrastructure isn't there for Overlay, passing the animation controller like this seems reasonable to me, especially because of its similarity to transitionBuilder's AnimationController parameters.

I'm all ears if @LongCatIsLooong or someone else has a better idea though.

Reply:
Replace the animation controller with a callback that returns a `Future`? Or two callbacks one for fade in and one for fade out? Right now people can change the animationController in unexpected ways. Same for `overlayEntry`.

Reply:
Wouldn't we still need a setter for this? the magnifier would need to take a controller and give it a callback to call when something calls `animationController.hide`. 

The fundamental problem here is that the magnifier does not exist when the the controller is created. The controller has no knowledge of what comes out of the builder, so it has no idea how long it's exit animation will take. Somehow, the widget that comes out of the builder needs to tell the controller that it's safe to remove from the overlay. I guess what you're asking is to have this (animationController)  passed in at the builders constructor time?
 but I don't think that's possible, since we don't want the loupe around at the controller's constructor time.

One solution that I can think of is have MagnifierConfiguration also have an `exitAnimationDuration`, and then the controller waits that duration to remove from the overlay. My problem with that is that it introduces a potential race condition and a weird API where the magnifier is expected to adhere to a duration contract, or it will be removed early / late.

Maybe we should schedule a call and discuss this in more depth?

Reply:
but overlay is a good call, I made it private